src/algorithm/greedy_algorithm.py

```diff
@@ -1,30 +1,40 @@
 import sys
 
 from algorithm.result import Result
-from problem.cvrp import Cvrp
 from problem.cvrp import cost
-from utils.configuration import Config
 
+from src.algorithm.algorithm import Algorithm
 
-def find_nearest(cvrp: Cvrp, city_id: int, visited_cities: list[int]) -> int:
-    closest = sys.maxsize
-    closest_index = 0
-    for i, value in enumerate(cvrp.distances_matrix[city_id]):
-        if i not in visited_cities and value < closest:
-            closest_index = i
-            closest = value
-    return closest_index
 
+class GreedyAlgorithm(Algorithm):
+    _current_city: int
+    _visited_cities: list[int]
 
-def solve_cvrp_greedy(cvrp: Cvrp, config: Config) -> Result:
-    current_city_id = 0
-    solution = [current_city_id]
-    for _ in range(1, cvrp.no_of_cities):
-        nearest_city_id = find_nearest(cvrp, current_city_id, solution)
-        solution.append(nearest_city_id)
-        current_city_id = nearest_city_id
+    def _initialize_algorithm(self) -> None:
+        self._visited_cities = []
+        self.visit_city(0)
 
-    best = average = cost(cvrp, solution)
-    best_genotype = solution
+    @property
+    def current_city(self) -> int:
+        return self._current_city
 
-    return Result(best, round(average, 2), best_genotype)
+    @property
+    def visited_cities(self) -> list[int]:
+        return self._visited_cities
+
+    def visit_city(self, city_id: int) -> None:
+        self._current_city = city_id
+        self._visited_cities.append(city_id)
+
+    def find_nearest_city_not_yet_visited(self) -> int:
+        return min(range(self.no_of_cities), key=lambda city: self.distances_matrix[self.current_city][
```

Comment:
It's O(n^2) - checks distance from current_city to each other city and for each check, it makes sure that the city has not been visited yet (need to iterate through the list again).
I will fix it.

Reply:
Visited cities could be a set, this could help with checking for duplicates.

Reply:
We might want to have a "solution" list that all algorithms have and a visisted set that's specific to this one?

Reply:
For greedy algorithm I'm representing the solution as dictionary (internally) in the end.